src/audio/google/google_rtc_audio_processing.c

```diff
@@ -544,222 +546,311 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
 	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
 	struct list_item *source_buffer_list_item;
 	struct comp_buffer *output;
-	unsigned int aec_channels = 0, frame_fmt, rate;
-	int microphone_stream_channels = 0;
-	int output_stream_channels;
-	int ret;
-	int i = 0;
+	int ret = 0, i = 0;
 
 	comp_info(dev, "google_rtc_audio_processing_prepare()");
 
-#if CONFIG_IPC_MAJOR_4
-	google_rtc_audio_processing_params(mod);
-#endif
-
 	/* searching for stream and feedback source buffers */
 	list_for_item(source_buffer_list_item, &dev->bsource_list) {
 		struct comp_buffer *source = container_of(source_buffer_list_item,
 							  struct comp_buffer, sink_list);
-#if CONFIG_IPC_MAJOR_4
-		if (IPC4_SINK_QUEUE_ID(buf_get_id(source)) ==
-			SOF_AEC_FEEDBACK_QUEUE_ID) {
-#else
-		if (source->source->pipeline->pipeline_id != dev->pipeline->pipeline_id) {
-#endif
+		if (is_ref_buffer(dev, source)) {
 			cd->aec_reference_source = i;
-			aec_channels = audio_stream_get_channels(&source->stream);
-			comp_dbg(dev, "reference index = %d, channels = %d", i, aec_channels);
+			cd->ref_comp_buffer = source;
 		} else {
 			cd->raw_microphone_source = i;
-			microphone_stream_channels = audio_stream_get_channels(&source->stream);
-			comp_dbg(dev, "microphone index = %d, channels = %d", i,
-				 microphone_stream_channels);
 		}
-
-		audio_stream_init_alignment_constants(1, 1, &source->stream);
 		i++;
 	}
-
 	output = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
 
-	/* On some platform the playback output is left right left right due to a crossover
-	 * later on the signal processing chain. That makes the aec_reference be 4 channels
-	 * and the AEC should only use the 2 first.
+	/* Validate channel, format and rate on each of our three
+	 * inputs.  All much match our build-time configuration, AEC
+	 * does not handle dynamic stream formats.
 	 */
-	if (cd->num_aec_reference_channels > aec_channels) {
-		comp_err(dev, "unsupported number of AEC reference channels: %d",
-			 aec_channels);
-		return -EINVAL;
+	int ref_fmt = source_get_frm_fmt(sources[cd->aec_reference_source]);
+	int ref_chan = source_get_channels(sources[cd->aec_reference_source]);
+	int ref_rate = source_get_rate(sources[cd->aec_reference_source]);
+
+	int mic_fmt = source_get_frm_fmt(sources[cd->raw_microphone_source]);
+	int mic_chan = source_get_channels(sources[cd->raw_microphone_source]);
+	int mic_rate = source_get_rate(sources[cd->raw_microphone_source]);
+
+	int out_fmt = sink_get_frm_fmt(sinks[0]);
+	int out_chan = sink_get_channels(sinks[0]);
+	int out_rate = sink_get_rate(sinks[0]);
+
+	/* Too many channels is a soft failure, AEC treats only the first N */
+	if (ref_chan > REF_CHAN_MAX)
```

Comment:
I don't see the truncation ... maybe this is intentional.

Reply:
Wouldn't it make sense to do the truncation here for clarity and affect it the to the state:

`state->num_reference_channels = MIN(REF_CHANNEL_MAX, requested_ref_channel)`

So we won't need extra logic in the copy routine.